spec/firebase_app_distribution_auth_client_spec.rb

```diff
@@ -29,87 +29,191 @@
-    end
+      it 'crashes if the service credentials file is not found' do
+        expect(File).to receive(:open)
+          .with("invalid_service_path")
+          .and_raise(Errno::ENOENT.new("file not found"))
+        expect { auth_client.fetch_auth_token("invalid_service_path", nil) }
+          .to raise_error("#{ErrorMessage::SERVICE_CREDENTIALS_NOT_FOUND}: invalid_service_path")
+      end
 
-    it 'uses firebase tools json if there is not another auth method' do
-      allow(File).to receive(:read)
-        .and_return(fake_firebase_tools_contents)
-      expect(File).to receive(:exist?).and_return(true)
-      expect(auth_helper.fetch_auth_token(nil)).to eq("fake_auth_token")
-    end
+      it 'crashes if the service credentials are invalid' do
+        expect(fake_service_creds).to receive(:fetch_access_token!)
+          .and_raise(Signet::AuthorizationError.new("error"))
+        expect { auth_client.fetch_auth_token("invalid_service_path", nil) }
+          .to raise_error("#{ErrorMessage::SERVICE_CREDENTIALS_ERROR}: invalid_service_path")
+      end
+
+      it 'crashes if given an invalid firebase token' do
+        expect(firebase_auth).to receive(:new)
+          .and_raise(Signet::AuthorizationError.new("error"))
+        expect { auth_client.fetch_auth_token(nil, "invalid_refresh_token") }
+          .to raise_error("#{ErrorMessage::REFRESH_TOKEN_ERROR}: invalid_refresh_token")
+      end
+
+      it 'crashes if the firebase tools has no tokens field' do
+        allow(File).to receive(:read)
+          .and_return(fake_firebase_tools_contents_no_tokens_field)
+        expect(File).to receive(:exist?).and_return(true)
+        expect { auth_client.fetch_auth_token(nil, nil) }
+          .to raise_error(ErrorMessage::MISSING_CREDENTIALS)
+      end
 
-    it 'fails if the firebase tools has no tokens field' do
-      allow(File).to receive(:read)
-        .and_return(fake_firebase_tools_contents_no_tokens_field)
-      expect(File).to receive(:exist?).and_return(true)
-      expect { auth_helper.fetch_auth_token(nil) }
-        .to raise_error(ErrorMessage::MISSING_CREDENTIALS)
+      it 'crashes if the firebase tools has no refresh_token field' do
+        allow(File).to receive(:read)
+          .and_return(fake_firebase_tools_contents_no_refresh_field)
+        expect(File).to receive(:exist?).and_return(true)
+        expect { auth_client.fetch_auth_token(nil, nil) }
+          .to raise_error(ErrorMessage::MISSING_CREDENTIALS)
+      end
     end
 
-    it 'fails if the firebase tools has no refresh_token field' do
-      allow(File).to receive(:read)
-        .and_return(fake_firebase_tools_contents_no_refresh_field)
-      expect(File).to receive(:exist?).and_return(true)
-      expect { auth_helper.fetch_auth_token(nil) }
-        .to raise_error(ErrorMessage::MISSING_CREDENTIALS)
+    describe 'with all other auth variables as empty' do
```

Comment:
Good call to test both the empty and nil cases. I think if you can avoid essentially copying all the tests from above, though, it would be good. Having everything twice makes it harder to maintain these tests. I think there are a couple good alternative options:

- Wrap your tests in a loop that will test one of multiple values, in this case nil and empty ([example](https://github.com/rspec/rspec-core/issues/2663#issuecomment-533634878))
- If you can, select a subset of the tests that will exercise enough of the code to check the nil vs. "" case. For example, would a single test that just ran `auth_client.fetch_auth_token("", "")` and asserted that it raises a `MISSING_CREDENTIALS` be enough to test that the code respects `""` as a nonexistent value?

Reply:
Thanks! I was able to wrap all the specs in a loop, exactly what I was looking for